File: CS/TaskHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace DxSample {
    public class TaskHelper : IDisposable {
        CancellationTokenSource tokenSource;
        CancellationToken cancellationToken;

        public TaskHelper() {
            this.tokenSource = new CancellationTokenSource();
            this.cancellationToken = tokenSource.Token;
        }
        public Task<string> StartTask(IProgress<int> progress) {
            return StartTaskCore(progress);
        }
        public void Cancel() {
            tokenSource.Cancel();
        }

        Task<string> StartTaskCore(IProgress<int> progress) {
            return Task.Run(() => CalculateResult(progress), cancellationToken);
        }
        string CalculateResult(IProgress<int> progress) {
            for(int n = 0; n <= 100; n++) {
                // do something helpful
                Thread.Sleep(60);

                cancellationToken.ThrowIfCancellationRequested();
                if(progress != null) progress.Report(n);
            }
            return "CalculatedValue";
        }

        #region IDisposable

        public void Dispose() {
            Dispose(true);
        }
        protected virtual void Dispose(bool disposing) {
            if(disposing) {
                if(tokenSource != null) tokenSource.Dispose();
            }
            this.tokenSource = null;
        }

        #endregion
    }
}

File: VB/TaskHelper.vb
Imports System
Imports System.Threading
Imports System.Threading.Tasks

Namespace DxSample

    Public Class TaskHelper
        Implements IDisposable

        Private tokenSource As CancellationTokenSource

        Private cancellationToken As CancellationToken

        Public Sub New()
            tokenSource = New CancellationTokenSource()
            cancellationToken = tokenSource.Token
        End Sub

        Public Function StartTask(ByVal progress As IProgress(Of Integer)) As Task(Of String)
            Return StartTaskCore(progress)
        End Function

        Public Sub Cancel()
            tokenSource.Cancel()
        End Sub

        Private Function StartTaskCore(ByVal progress As IProgress(Of Integer)) As Task(Of String)
            Return Task.Run(Function() CalculateResult(progress), cancellationToken)
        End Function

        Private Function CalculateResult(ByVal progress As IProgress(Of Integer)) As String
            For n As Integer = 0 To 100
                ' do something helpful
                Thread.Sleep(60)
                cancellationToken.ThrowIfCancellationRequested()
                If progress IsNot Nothing Then progress.Report(n)
            Next

            Return "CalculatedValue"
        End Function

#Region "IDisposable"
        Public Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
        End Sub

        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If tokenSource IsNot Nothing Then tokenSource.Dispose()
            End If

            tokenSource = Nothing
        End Sub
#End Region
    End Class
End Namespace
